session/session.go

```diff
@@ -1411,6 +1414,29 @@ func (s *session) ExecuteStmt(ctx context.Context, stmtNode ast.StmtNode) (sqlex
 	return recordSet, nil
 }
 
+func (s *session) validateStatementReadOnlyInStaleness(stmtNode ast.StmtNode) error {
+	vars := s.GetSessionVars()
+	if !vars.TxnCtx.IsStaleness {
+		return nil
+	}
+	errMsg := "only support read-only statement during read-only staleness transactions"
+	switch stmtNode.(type) {
+	case *ast.SplitRegionStmt:
+		return nil
+	case *ast.SelectStmt, *ast.ExecuteStmt, *ast.ExplainStmt, *ast.ShowStmt, *ast.SetOprStmt:
+		// covered by IsReadOnly
+		if !planner.IsReadOnly(stmtNode, vars) {
```

Comment:
`planner.IsReadOnly` covers `ExecuteStmt`. Then It invokes `parser/ast.IsReadOnly` to cover `Select, Explain, Show, Do, SetOpr, SetOprSelectList`.
But this switch case does not contain `Do` & `SetOprSelectList`.

I think we better use whitelist, I mean a structure like:
```
if !vars.TxnCtx.IsStaleness ...
if planner.IsReadOnly ... // allow readonly stmts
// allow SplitRegion and other stmts that is not included in planner.IsReadOnly
// forbid all other stmts
```